File: trees/trie.py
# -*- coding: utf-8 -*-
# @Author: krocki
# @Date:   2017-01-12 20:22:32
# @Last Modified by:   krocki
# @Last Modified time: 2017-01-12 21:16:30

def find(nodes, val):

	l = [i for i in nodes if i.val == val]
	if l: return l[0]
	else: return None

# TODO: improve
def traverse(node):

	print "{" + str([traverse(i) for i in node.children]) + "}"

	return node.val

def lookup(node, string):
	
	if len(string) == 0:
		return node
	else:
		x = string[0]
		n = find(node.children, x)
		if n == None: return None
		else: 
			if len(string) == 1: 
				traverse(node)
				return n
			else: return lookup(n, string[1:])

def insert(node, string):

	if len(string) == 0: x = '*'
	else: x = string[0]

	n = find(node.children, x)

	if n == None:
		n = Node(x)
		node.children.append(n)

	if len(string) > 0:
		xs = string[1:]
		insert(n, xs)
	

class Node:

	children = []
	val = None

	def __init__(self, val):
		self.val = val
		self.children = []

root = Node("#")
insert(root, "car")
insert(root, "cat")
insert(root, "cart")
insert(root, "cars")

traverse(root)

print "looking up... ca"
lookup(root, "ca")

File: trees/trie.hs
-- @Author: krocki
-- @Date:   2017-01-11 16:10:29
-- @Last Modified by:   krocki
-- @Last Modified time: 2017-01-12 16:14:33

-- We're going to make our own Contacts
-- application! The application must perform two
-- types of operations:

-- add name, where  is a string denoting a contact
-- name. This must store  as a new contact in the
-- application. find partial, where  is a string
-- denoting a partial name to search the
-- application for. It must count the number of
-- contacts starting with  and print the count on
-- a new line.

data Trie = Node Char [Trie] deriving (Eq, Show)

-- returns the modified list of Tries
insert (x:xs) [] = [Node x (insert xs [])]
insert [] [] = [(Node '*' [])]
insert [] ((Node c' l'):ls) = ((Node c' (insert "*" l')):ls)
insert xx@(x:xs) (l@(Node c' l'):ls)
                        | x == c' = ((Node c' (insert xs l')):ls)
                        | x /= c' = (l:insert xx ls)

add :: String -> Trie -> Trie
add xx (Node c l) = (Node c (insert xx l))

main = do
    -- root is marked by symbol "#", word ending is marked by "*"
    let root = add "cars" (add "cat" (add "car" (Node '#' [])))
    print root
    -- == Node '#' 
        -- [Node 'c' 
                -- [Node 'a' 
                    -- [Node 'r' 
                        -- [Node '*' [], Node 's' 
                                -- [Node '*' []
                                -- ]
                        -- ],
                    -- Node 't' 
                        -- [Node '*' []
                        -- ]
                    -- ]
                -- ]
        -- ]

